src/overlays/kaleido_scope/ovl_kaleido_scope/z_kaleido_map.c

```diff
@@ -1,108 +1,945 @@
+    s16 rectRight;
+
+    OPEN_DISPS(play->state.gfxCtx);
+
+    KaleidoScope_SetCursorVtx(pauseCtx, pauseCtx->cursorSlot[PAUSE_MAP] * 4, pauseCtx->mapPageVtx);
+
+    if ((pauseCtx->pageIndex == PAUSE_MAP) && (pauseCtx->state == 6) &&
+        ((pauseCtx->unk_200 == 0) || (pauseCtx->unk_200 == 3)) && YREG(6) && (pauseCtx->state != 7) &&
+        !((pauseCtx->state >= 8) && (pauseCtx->state <= 0x12))) {
+
+        func_8012C628(play->state.gfxCtx);
+
+        gDPSetTextureFilter(POLY_OPA_DISP++, G_TF_POINT);
+        gDPLoadTLUT_pal256(POLY_OPA_DISP++, gWorldMapImageTLUT);
+        gDPSetTextureLUT(POLY_OPA_DISP++, G_TT_RGBA16);
+
+        gDPSetPrimColor(POLY_OPA_DISP++, 0, 0, 255, 255, 255, pauseCtx->alpha);
+
+        // Process all 128 rows of pixels for gWorldMapImageTex over 16 iterations
+        // Loop over yPos (t), textureIndex (j)
+        for (t = 62, j = 0; j < 16; j++, t += 8) {
+            gDPLoadTextureBlock(POLY_OPA_DISP++, (u8*)gWorldMapImageTex + j * (WORLD_MAP_IMAGE_TEX_WIDTH * 8),
+                                G_IM_FMT_CI, G_IM_SIZ_8b, WORLD_MAP_IMAGE_TEX_WIDTH, 8, 0, G_TX_NOMIRROR | G_TX_WRAP,
+                                G_TX_NOMIRROR | G_TX_WRAP, G_TX_NOMASK, G_TX_NOMASK, G_TX_NOLOD, G_TX_NOLOD);
+
+            rectLeft = 51 << 2;
+            rectRight = rectLeft + (WORLD_MAP_IMAGE_TEX_WIDTH << 2);
+            gSPTextureRectangle(POLY_OPA_DISP++, rectLeft, t << 2, rectRight, (t << 2) + (8 << 2), G_TX_RENDERTILE, 0,
+                                0, 1 << 10, 1 << 10);
+        }
+
+        func_8012C8AC(play->state.gfxCtx);
+
+    } else {
+        gDPPipeSync(POLY_OPA_DISP++);
+        gDPSetTextureFilter(POLY_OPA_DISP++, G_TF_POINT);
+        gDPLoadTLUT_pal256(POLY_OPA_DISP++, gWorldMapImageTLUT);
+        gDPSetTextureLUT(POLY_OPA_DISP++, G_TT_RGBA16);
+
+        gDPSetPrimColor(POLY_OPA_DISP++, 0, 0, 255, 255, 255, pauseCtx->alpha);
+        gSPVertex(POLY_OPA_DISP++, &pauseCtx->mapPageVtx[204], 32, 0);
+
+        // Process the first 72 rows of pixels for gWorldMapImageTex over 8 iterations
+        // Loop over textureIndex (i, k), vtxIndex (j)
+        for (i = 0, k = 0, j = 0; i < 8; i++, k++, j += 4) {
+            gDPLoadTextureBlock(POLY_OPA_DISP++, (u8*)gWorldMapImageTex + k * (WORLD_MAP_IMAGE_TEX_WIDTH * 9),
+                                G_IM_FMT_CI, G_IM_SIZ_8b, WORLD_MAP_IMAGE_TEX_WIDTH, 9, 0, G_TX_NOMIRROR | G_TX_WRAP,
+                                G_TX_NOMIRROR | G_TX_WRAP, G_TX_NOMASK, G_TX_NOMASK, G_TX_NOLOD, G_TX_NOLOD);
+
+            gSP1Quadrangle(POLY_OPA_DISP++, j, j + 2, j + 3, j + 1, 0);
+        }
+
+        gSPVertex(POLY_OPA_DISP++, &pauseCtx->mapPageVtx[236], 28, 0);
+
+        // Process the next 54 rows of pixels for gWorldMapImageTex over 6 iterations
+        // Loop over textureIndex (i, k), vtxIndex (j)
+        for (i = 0, j = 0; i < 6; i++, k++, j += 4) {
+            gDPLoadTextureBlock(POLY_OPA_DISP++, (u8*)gWorldMapImageTex + k * (WORLD_MAP_IMAGE_TEX_WIDTH * 9),
```

Comment:
Just checked, the 9 makes sense: TMEM_SIZE is 0x1000. For a CI half is used for the palette, so you have 0x800 left. Floor of 0x800 / 216 is 9, as expected. So you could write this as `TMEM_SIZE / 2 / WIDTH`.

Reply:
I think it would have to be all or nothing with all the loops. For example, if you wanted to change the width of the image, then this for-loop would adjust, but then the other loops would be incorrect. Maybe just leave it as a comment as to where the 9 comes from?